model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py

```diff
@@ -121,3 +125,90 @@ def find_and_replace_pattern(self, graph: Graph):
             for visited_node in marked_nodes:
                 mark_as_correct_data_layout(visited_node)
                 visited_node['nchw_layout'] = True
+
+        for node in self.get_ports_and_nodes_on_weights(graph)[1]:
+            mark_as_correct_data_layout(node)
+            node['nchw_layout'] = True
+            if node.soft_get('type') == 'Const':  # WA for Const op deletion during clean_up
+                node.out_node()['nchw_layout'] = True
+
+        for node in self.get_ports_and_nodes_on_shape_subgraphs(graph)[1]:
+            mark_as_correct_data_layout(node)
+            node['nchw_layout'] = True
+            if node.soft_get('type') == 'Const':  # WA for Const op deletion during clean_up
+                node.out_node()['nchw_layout'] = True
+
+    @staticmethod
+    def walk_up_from_in_ports_to_out_ports(in_ports: Set[Port], out_ports: Set[Port],
+                                           visited_ports: Set[Port] = None, visited_nodes: Set[Node] = None):
+        """"
+        Returns all intermediate ports and nodes of such a sub-graph:
+
+            out_ports
+            |        |
+           \/       \/
+            .   .   .
+            |        |
+           \/       \/
+            in_ports
+        """
+        if visited_ports is None:
+            visited_ports = set()
+        if visited_nodes is None:
+            visited_nodes = set()
+
+        deque_of_in_ports = deque(in_ports)
+        while len(deque_of_in_ports):
+            in_port = deque_of_in_ports.popleft()
+            source_node = in_port.get_source().node
+            if in_port in visited_ports or source_node in visited_nodes:
```

Comment:
When you visit some `source_node` you add it to the `visited_nodes`. But you add just one input port `in_port` of the `source_node` to `visited_ports`, because here you check that the `source_node` has been visited already and if so then do not continue, while the `in_port` is added to `visited_ports` in line 166. Is it expected behaviour?

Reply:
In the code above (`find_and_replace_pattern` function) you use only nodes and not ports. So may be remove `visited_ports` at all?

Reply:
I was into collecting all the graph pieces from such a sub-graphs

![image](https://user-images.githubusercontent.com/55839243/93474455-06587c80-f900-11ea-9ef1-b77efef12217.jpeg)

Reply:
It doesn't cost much to support but it would be extremely useful in case we consider solving layout propagation.